compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/codegen/SwitchGenerator.kt

```diff
@@ -137,58 +173,190 @@ class SwitchGenerator(private val expression: IrWhen, private val data: BlockInf
     private fun coerceNotToUnit(fromType: Type, fromKotlinType: KotlinType?, toKotlinType: KotlinType): StackValue =
         codegen.coerceNotToUnit(fromType, fromKotlinType, toKotlinType)
 
-    private fun gen(
-        cases: ArrayList<Pair<Int, Label>>,
+    abstract inner class Switch(
+        val subject: IrGetValue,
+        val defaultLabel: Label,
+        val endLabel: Label,
+        val elseExpression: IrExpression?,
+        val thenExpressions: ArrayList<Pair<IrExpression, Label>>
+    ) {
+        open fun gen(): StackValue? {
```

Comment:
I would prefer to explicitly clarify here that gen works only if there are enough cases
gen -> genOptimizedIfEnoughCases (not sure that new name couldn't be better)?